### PR TITLE
feat: implement a regression test for large uploads

### PR DESCRIPTION
**Pull request by coryan**
This is a regression test, but so far does not reproduce a problem. But
it might help us diagnose the problem reported in #3029 

The test runs multiple threads, all of them uploading a mid-sized
object (128MiB). The test uses `InsertObject()` and `WriteObject()`
to perform the upload, and in the case of `WriteObject()` it both
uploads the object using a single `operator<<` call, and multiple
`.write()` calls.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/3066)

### CODE REVIEW

File: google/cloud/storage/tests/parallel_upload_regression/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+RUN apt update && \
+    apt install -y libstdc++6 libc-ares2 libcurl4 zlib1g binutils \
+        ca-certificates tcpdump coreutils strace procps
+RUN /usr/sbin/update-ca-certificates
+
+COPY --from=build \
+    /w/.build/build/google-cloud-cpp/src/google_cloud_cpp_project-build/google/cloud/storage/tests/parallel_upload_regression/parallel_upload_test \
+    /r/
+
+CMD ["/bin/false"]
```
Comment (devbww) [inline google/cloud/storage/tests/parallel_upload_regression/Dockerfile]: What does this do?

File: google/cloud/storage/tests/parallel_upload_regression/parallel_upload_test.cc [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+      auto writer = client.WriteObject(flag_dst_bucket_name, name);
+      for (std::size_t offset = 0; offset < upload_contents_.size();
+           offset += chunk_size) {
+        auto count = (std::min)(chunk_size, upload_contents_.size() - offset);
+        writer.write(upload_contents_.data() + offset, count);
+        EXPECT_FALSE(writer.bad());
+        EXPECT_STATUS_OK(writer.last_status());
+      }
+      writer.Close();
+      std::cout << '.' << std::flush;
```
Comment (devbww) [inline google/cloud/storage/tests/parallel_upload_regression/parallel_upload_test.cc]: When do these lines get terminated?

Comment (coryan) [reply google/cloud/storage/tests/parallel_upload_regression/parallel_upload_test.cc]: Thanks, fixed.

File: google/cloud/storage/tests/parallel_upload_regression/Dockerfile (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/tests/parallel_upload_regression/Dockerfile]: Set the default command for `docker run` to `/bin/false`.  Harder to misuse the resulting docker image that way (I think).

Comment (devbww) [reply google/cloud/storage/tests/parallel_upload_regression/Dockerfile]: Okie dokie.  Thanks.

**Pull request merged**